tasks_test.go

```diff
@@ -0,0 +1,180 @@
+		}
+		break
+	}
+	return nil
+}
+
+func awaitTaskRunning(client *Client, taskID string) error {
+	for range time.Tick(100 * time.Millisecond) {
+		res, err := client.Tasks().GetStatus(taskID)
+		if err != nil {
+			return err
+		}
+
+		if res.Status.Status == "RUNNING" {
+			return nil
+		}
+	}
+	return errors.New("task has not started")
+}
+
+func runInlineIngestionTask(client *Client, dataSourceName string, entries []TestDO, recordsCount int) error {
+	taskID, err := triggerIngestionTask(client, dataSourceName, entries)
+	if err != nil {
+		return err
+	}
+
+	err = awaitTaskCompletion(client, taskID)
+	if err != nil {
+		return err
+	}
+
+	err = client.Metadata().AwaitDataSourceAvailable(dataSourceName)
+	if err != nil {
+		return err
+	}
+
+	err = client.Metadata().AwaitRecordsCount(dataSourceName, recordsCount)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func TestTaskService(t *testing.T) {
+	// Set up druid containers using docker-compose.
+	compose, err := tc.NewDockerCompose("testdata/docker-compose.yaml")
+	require.NoError(t, err, "NewDockerComposeAPI()")
+
+	// Set up cleanup for druid containers.
+	t.Cleanup(func() {
+		require.NoError(t, compose.Down(context.Background(), tc.RemoveOrphans(true), tc.RemoveVolumes(true), tc.RemoveImagesLocal), "compose.Down()")
+	})
+
+	ctx, cancel := context.WithCancel(context.Background())
+	t.Cleanup(cancel)
+
+	// Set up druid service and client.
+	var druidOpts []ClientOption
+	d, err := NewClient("http://localhost:8888", druidOpts...)
+	require.NoError(t, err, "error should be nil")
+
+	// Waiting for druid services to start.
+	err = compose.
+		WaitForService("coordinator", wait.NewHTTPStrategy(processInformationPathPrefix).WithPort("8081/tcp").WithStartupTimeout(180*time.Second)).
+		WaitForService("router", wait.NewHTTPStrategy(processInformationPathPrefix).WithPort("8888/tcp").WithStartupTimeout(180*time.Second)).
+		WaitForService("broker", wait.NewHTTPStrategy(processInformationPathPrefix).WithPort("8082/tcp").WithStartupTimeout(180*time.Second)).
+		WaitForService("middlemanager", wait.NewHTTPStrategy(processInformationPathPrefix).WithPort("8091/tcp").WithStartupTimeout(180*time.Second)).
+		Up(ctx, tc.Wait(true))
+	require.NoError(t, err, "druid services should be up with no error")
+
+	// Test create ingestion task -> get status -> complete sequence.
+	runInlineIngestionTask(d, "test-submit-task-datasource", testObjects, 2)
+	require.NoError(t, err, "error should be nil")
+}
+
+func TestTerminateTask(t *testing.T) {
+	// Set up druid containers using docker-compose.
+	compose, err := tc.NewDockerCompose("testdata/docker-compose.yaml")
+	require.NoError(t, err, "NewDockerComposeAPI()")
+
+	// Set up cleanup for druid containers.
+	t.Cleanup(func() {
+		require.NoError(t, compose.Down(context.Background(), tc.RemoveOrphans(true), tc.RemoveVolumes(true), tc.RemoveImagesLocal), "compose.Down()")
+	})
+
+	ctx, cancel := context.WithCancel(context.Background())
+	t.Cleanup(cancel)
```

Comment:
Can you explain? 

Why `Cleanup` and not `defer`?
Why even cancel when it's not being called explicitly beyond the cleanup?

Reply:
Fixed.